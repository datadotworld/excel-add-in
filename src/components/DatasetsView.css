/*
 * Copyright 2017 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc. (http://data.world/).
 */
.datasets-view .dataset-selector .section-header {
  padding: 15px;
  display: flex;
  justify-content: space-between;
  align-items: baseline;
}

.datasets-view .dataset-selector .section-header .add-icon {
  width: 15px;
  height: 15px;
  margin-right: 5px;
  cursor: pointer;
}

.datasets-view .dataset-selector .section-header .add-icon .stroke {
  stroke: #4376b2;
}

.datasets-view .section-header .title {
  font-size: 18px;
  font-weight: 700;
}

.datasets-view .loaderoverlay {
  position: inherit;
  top: initial;
  bottom: initial;
  left: initial;
  right: initial;
}

.datasets-view .dataset-selector .section-header button {
  border: none;
  box-shadow: none;
  padding: 0;
  color: #4376b2;
  float: right;
  font-size: 18px;
}