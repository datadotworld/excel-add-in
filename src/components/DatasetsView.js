  /*
 * Copyright 2017 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc. (http://data.world/).
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { 
  Button,
  Grid,
  Row,
  ControlLabel
} from 'react-bootstrap';
import './DatasetsView.css';
import DatasetItem from './DatasetItem';
import LoadingAnimation from './LoadingAnimation';
import analytics from '../analytics';

class DatasetsView extends Component {

  static propTypes = {
    createDataset: PropTypes.func,
    datasets: PropTypes.array,
    linkDataset: PropTypes.func,
    loadingDatasets: PropTypes.bool
  }

  static defaultProps = {
    datasets: [],
    loadingDatasets: true
  }

  state = {
    sortKey: 'updated'
  }

  sortDatasets = () => {
    const sortKey = this.state.sortKey;
    const sortedDatasets = this.props.datasets.slice();
    const reverseSort = sortKey.indexOf('-') === 0;

    sortedDatasets.sort((a, b) => {
      if (sortKey.indexOf('title') >= 0) {
        if (a.title < b.title) {
          return reverseSort ?  1 : -1;
        } else if (a.title > b.title) {
          return reverseSort ? -1 : 1;
        }
        return 0;
      } else {
        let dateA, dateB;
        if (sortKey.indexOf('updated') >= 0) {
          dateA = new Date(a.updated);
          dateB = new Date(b.updated);
        } else {
          dateA = new Date(a.created);
          dateB = new Date(b.created);
        }
        return reverseSort ? dateA - dateB : dateB - dateA;
      }
    });
    return sortedDatasets;
  }

  sortChanged = (sortKey) => {
    analytics.track('exceladdin.datasets.sort.change', {sort: sortKey});
    this.setState({sortKey})
  }

  createDatasetClick = () => {
    analytics.track('exceladdin.datasets.create_dataset_button.click');
    this.props.createDataset();
  }

  addDatasetClick = () => {
    analytics.track('exceladdin.datasets.create_dataset_add.click');
    this.props.createDataset();
  }

  render () {
    const { datasets, loadingDatasets, showCreateDataset, linkDataset } = this.props;
    const sortedDatasets = this.sortDatasets();
    const datasetEntries = sortedDatasets.map((d) =>{
      return (<DatasetItem dataset={d} key={`${d.owner}/${d.id}`} buttonText='Link' buttonHandler={linkDataset} />);
    });

    return (
      <Grid className='datasets-view'>
        <div className='dataset-selector'>
          <Row className='center-block section-header'>
            <ControlLabel>
              Select a dataset or project
            </ControlLabel>
            <Button bsStyle='default' onClick={() => this.props.showDatasets()}>Cancel</Button>
          </Row>
          {loadingDatasets && <LoadingAnimation label='Fetching datasets...' />}
          {!!datasets.length && !loadingDatasets &&
            <Row className='center-block'>
              <div>
                <DatasetItem buttonText='Link' dataset={{isCreate: true}} buttonHandler={showCreateDataset}/>
                {datasetEntries}
              </div>
            </Row>}
          {!datasets.length && !loadingDatasets &&
            <Row className='center-block no-datasets'>
              <div className='message'>
                You haven't created any datasets to link data to.
              </div>
              <Button className='bottom-button' bsStyle='primary' onClick={this.createDatasetClick}>Create a new dataset</Button>
            </Row>
          }
        </div>
      </Grid>
    );
  }
}

export default DatasetsView;