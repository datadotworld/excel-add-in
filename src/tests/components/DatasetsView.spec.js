/*
 * Copyright 2017 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc. (http://data.world/).
 */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { IntlProvider } from 'react-intl';

import DatasetsView from '../../components/DatasetsView';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<DatasetsView />, div);
});

it('renders datasets view - loading state', () => {
  expect(renderer.create(<DatasetsView loadingDatasets />).toJSON()).toMatchSnapshot()
});

it('renders datasets view - no datasets', () => {
  expect(renderer.create(<DatasetsView loadingDatasets={false} />).toJSON()).toMatchSnapshot()
});

it('renders datasets view - datasets', () => {
  const datasets = [
    {
      owner: 'test',
      id: 'test',
      title: 'This is a test',
      updated: '2017-07-20T14:24:51.762Z',
      created: '2017-07-20T14:24:51.762Z'
    },
    {
      owner: 'test',
      id: 'test2',
      title: 'This is a test too',
      updated: '2017-08-20T14:24:51.762Z',
      created: '2017-08-20T14:24:51.762Z'
    },
    {
      owner: 'test',
      id: 'test3',
      title: 'This is a test of a project',
      updated: '2017-08-20T14:24:51.762Z',
      created: '2017-08-20T14:24:51.762Z',
      isProject: true
    }
  ];

  expect(renderer.create(
    <IntlProvider locale='en'>
      <DatasetsView datasets={datasets} loadingDatasets={false} />
    </IntlProvider>).toJSON()).toMatchSnapshot()
});