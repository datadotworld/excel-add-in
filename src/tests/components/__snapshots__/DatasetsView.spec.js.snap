// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders datasets view - datasets 1`] = `
<div
  className="datasets-view container"
>
  <div
    className="dataset-selector"
  >
    <div
      className="center-block section-header row"
    >
      <label
        className="control-label"
        htmlFor={undefined}
      >
        Select a dataset or project
      </label>
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
    <div
      className="center-block row"
    >
      <div>
        <div
          className="dataset"
          onClick={[Function]}
        >
          <img
            className="add-icon"
            src="icon-add.svg"
          />
          <div
            className="center-info"
          >
            <div
              className="title"
            >
              Create a new dataset
            </div>
          </div>
          <div
            className="link-icon"
          >
            <span
              className="svg-icon "
              onClick={undefined}
            >
              <svg
                className="angle-right"
                height="40"
                viewBox="0 0 40 40"
                width="40"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  className="stroke"
                  d="m26.3 21.4q0 0.3-0.2 0.5l-10.4 10.4q-0.3 0.3-0.6 0.3t-0.5-0.3l-1.1-1.1q-0.2-0.2-0.2-0.5t0.2-0.5l8.8-8.8-8.8-8.7q-0.2-0.3-0.2-0.6t0.2-0.5l1.1-1.1q0.3-0.2 0.5-0.2t0.6 0.2l10.4 10.4q0.2 0.2 0.2 0.5z"
                />
              </svg>
            </span>
          </div>
        </div>
        <div
          className="dataset"
          onClick={[Function]}
        >
          <span
            className="svg-icon "
            onClick={undefined}
          >
            <svg
              className="dataset-schema-icon"
              viewBox="0 0 16 16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                className="fill"
                d="M15,0H1C0.4,0,0,0.4,0,1v14c0,0.6,0.4,1,1,1h14c0.6,0,1-0.4,1-1V1C16,0.4,15.6,0,15,0z M8,13.5L2.2,9.9l1.5-1L8,11.5l4.3-2.7l1.5,0.9L8,13.5z M8,9.9L2.2,6.1L8,2.5l5.8,3.6L8,9.9z"
              />
            </svg>
          </span>
          <div
            className="center-info"
          >
            <div
              className="title"
            >
              This is a test too
            </div>
            <div
              className="info"
            >
              @
              test
               · Updated 
              <span>
                Aug 20, 2017
              </span>
            </div>
          </div>
          <div
            className="link-icon"
          >
            <span
              className="svg-icon "
              onClick={undefined}
            >
              <svg
                className="angle-right"
                height="40"
                viewBox="0 0 40 40"
                width="40"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  className="stroke"
                  d="m26.3 21.4q0 0.3-0.2 0.5l-10.4 10.4q-0.3 0.3-0.6 0.3t-0.5-0.3l-1.1-1.1q-0.2-0.2-0.2-0.5t0.2-0.5l8.8-8.8-8.8-8.7q-0.2-0.3-0.2-0.6t0.2-0.5l1.1-1.1q0.3-0.2 0.5-0.2t0.6 0.2l10.4 10.4q0.2 0.2 0.2 0.5z"
                />
              </svg>
            </span>
          </div>
        </div>
        <div
          className="dataset"
          onClick={[Function]}
        >
          <span
            className="svg-icon "
            onClick={undefined}
          >
            <svg
              className="project-schema-icon"
              version="1.1"
              viewBox="0 0 16 16"
              x="0px"
              xmlns="http://www.w3.org/2000/svg"
              y="0px"
            >
              <path
                className="fill"
                d="M15,0H1C0.4,0,0,0.4,0,1v14c0,0.6,0.4,1,1,1h14c0.6,0,1-0.4,1-1V1C16,0.4,15.6,0,15,0z M13.6,6.3l-2.3,3.5c-0.3,0.6-1.3,0.7-1.7,0.2L7,7.5l-2.7,3.6c-0.3,0.5-1.2,0.6-1.6,0.2C2.2,11,2.1,10.1,2.4,9.7L5.9,5c0.5-0.6,1.3-0.6,1.7-0.1l2.4,2.4L11.6,5c0.3-0.6,1-0.7,1.6-0.3C13.8,5,13.9,5.7,13.6,6.3z"
              />
            </svg>
          </span>
          <div
            className="center-info"
          >
            <div
              className="title"
            >
              This is a test of a project
            </div>
            <div
              className="info"
            >
              @
              test
               · Updated 
              <span>
                Aug 20, 2017
              </span>
            </div>
          </div>
          <div
            className="link-icon"
          >
            <span
              className="svg-icon "
              onClick={undefined}
            >
              <svg
                className="angle-right"
                height="40"
                viewBox="0 0 40 40"
                width="40"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  className="stroke"
                  d="m26.3 21.4q0 0.3-0.2 0.5l-10.4 10.4q-0.3 0.3-0.6 0.3t-0.5-0.3l-1.1-1.1q-0.2-0.2-0.2-0.5t0.2-0.5l8.8-8.8-8.8-8.7q-0.2-0.3-0.2-0.6t0.2-0.5l1.1-1.1q0.3-0.2 0.5-0.2t0.6 0.2l10.4 10.4q0.2 0.2 0.2 0.5z"
                />
              </svg>
            </span>
          </div>
        </div>
        <div
          className="dataset"
          onClick={[Function]}
        >
          <span
            className="svg-icon "
            onClick={undefined}
          >
            <svg
              className="dataset-schema-icon"
              viewBox="0 0 16 16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                className="fill"
                d="M15,0H1C0.4,0,0,0.4,0,1v14c0,0.6,0.4,1,1,1h14c0.6,0,1-0.4,1-1V1C16,0.4,15.6,0,15,0z M8,13.5L2.2,9.9l1.5-1L8,11.5l4.3-2.7l1.5,0.9L8,13.5z M8,9.9L2.2,6.1L8,2.5l5.8,3.6L8,9.9z"
              />
            </svg>
          </span>
          <div
            className="center-info"
          >
            <div
              className="title"
            >
              This is a test
            </div>
            <div
              className="info"
            >
              @
              test
               · Updated 
              <span>
                Jul 20, 2017
              </span>
            </div>
          </div>
          <div
            className="link-icon"
          >
            <span
              className="svg-icon "
              onClick={undefined}
            >
              <svg
                className="angle-right"
                height="40"
                viewBox="0 0 40 40"
                width="40"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  className="stroke"
                  d="m26.3 21.4q0 0.3-0.2 0.5l-10.4 10.4q-0.3 0.3-0.6 0.3t-0.5-0.3l-1.1-1.1q-0.2-0.2-0.2-0.5t0.2-0.5l8.8-8.8-8.8-8.7q-0.2-0.3-0.2-0.6t0.2-0.5l1.1-1.1q0.3-0.2 0.5-0.2t0.6 0.2l10.4 10.4q0.2 0.2 0.2 0.5z"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders datasets view - loading state 1`] = `
<div
  className="datasets-view container"
>
  <div
    className="dataset-selector"
  >
    <div
      className="center-block section-header row"
    >
      <label
        className="control-label"
        htmlFor={undefined}
      >
        Select a dataset or project
      </label>
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
    <div
      className="loaderoverlay"
    >
      <div
        className="loader-center"
      >
        <div
          className="loader"
        >
          Loading...
        </div>
        <span
          className="loader-label"
        >
          Fetching datasets...
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders datasets view - no datasets 1`] = `
<div
  className="datasets-view container"
>
  <div
    className="dataset-selector"
  >
    <div
      className="center-block section-header row"
    >
      <label
        className="control-label"
        htmlFor={undefined}
      >
        Select a dataset or project
      </label>
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
    <div
      className="center-block no-datasets row"
    >
      <div
        className="message"
      >
        You haven't created any datasets to link data to.
      </div>
      <button
        className="bottom-button btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Create a new dataset
      </button>
    </div>
  </div>
</div>
`;
